File: Copy_List_with_Random_Pointer.js
var copyRandomList = function(head) {
    const hashMap = new Map();
    let cur = head;

    while (cur) {
        hashMap.set(cur, new Node(cur.val));
        cur = cur.next;
    }

    cur = head;

    while (cur) {
        const copy = hashMap.get(cur);
        copy.next = hashMap.get(cur.next) || null;
        copy.random = hashMap.get(cur.random) || null;
        cur = cur.next;
    }

    return hashMap.get(head);   
};

File: Copy_List_with_Random_Pointer.java
class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> hashMap = new HashMap<>();
        Node cur = head;

        while (cur != null) {
            hashMap.put(cur, new Node(cur.val));
            cur = cur.next;
        }

        cur = head;

        while (cur != null) {
            Node copy = hashMap.get(cur);
            copy.next = hashMap.get(cur.next);
            copy.random = hashMap.get(cur.random);
            cur = cur.next;
        }

        return hashMap.get(head);        
    }
}
